File: 0525-contiguous-array/0525-contiguous-array.py
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        # p[k] = p[0] + p[1] + ... + p[i] + p[i+1] + ... + p[k]
        # sum(i+1, k) = p[k] - p[i]
        # hope sum(i+1, k) = 0
        # p[k] = p[i]

        diff = 0
        diff_index = {0:-1}
        res = 0

        for i, n in enumerate(nums):
            if n == 0:
                diff -= 1
            elif n == 1:
                diff += 1
            
            if diff in diff_index:
                res = max(res, i - diff_index[diff])
            else:
                diff_index[diff] = i
                
        return res

File: 0525-contiguous-array/0525-contiguous-array.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        unordered_map<int, int> m;
        int ans = 0;
        int cnt = 0;
        m[0] = -1;
        for(int i = 0; i < nums.size(); i++) {
            cnt += nums[i] ? 1 : -1;
            auto it = m.find(cnt);
            if(it != m.end()) ans = max(ans, i - it->second);
            else m[cnt] = i;
        }
        return ans;
    }
};